server/base/src/main/java/org/apache/accumulo/server/security/handler/ZKAuthenticator.java

```diff
@@ -52,6 +52,23 @@ public void initialize(ServerContext context) {
     this.context = context;
     zooCache = new ZooCache(context.getZooReaderWriter(), null);
     ZKUserPath = Constants.ZROOT + "/" + context.getInstanceID() + "/users";
+    checkOutdatedHashes();
+  }
+
+  private void checkOutdatedHashes() {
+    try {
+      listUsers().forEach(user -> {
+        String zpath = ZKUserPath + "/" + user;
+        byte[] zkData = zooCache.get(zpath);
+        if (ZKSecurityTool.isOutdatedPass(zkData)) {
+          log.warn("Found user(s) with outdated password hash. These will be re-hashed"
+              + " on successful authentication.");
+          return;
+        }
+      });
+    } catch (NullPointerException e) {
+      // initializeSecurity was not called yet, there could be no outdated passwords stored
```

Comment:
log a DEBUG message with these details.

Reply:
It looks like initialize itself is already an edge case... see my comment at https://github.com/apache/accumulo/pull/1798/files/4c8fedce90487c40ec4b740bc98dc181177cc71f#r528903518